### ISSUE TITLE
Check if dataset augmentation improves scores

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/auto_slide#16]
Dataset can be augmented by adding:

1. negative samples
2. artifical positive samples made by extracting labelled vessels, warping them, and adding them to negative samples

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **Identified Files for Modification:**
   - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/model/training.py`
     - **Description:** Contains the training logic for the model. Needs modification to include augmented datasets.
   - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/utils.py`
     - **Description:** Contains utility functions for data processing. Needs new functions for generating artificial positive samples.
   - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/run_pipeline.py`
     - **Description:** Orchestrates the complete pipeline. Needs to ensure the pipeline incorporates the augmented dataset.

2. **Plan to Address the Issues:**
   - **Modify Training Logic:** Update the training script to incorporate augmented datasets, including negative samples and artificially generated positive samples.
   - **Utility Functions for Augmentation:** Add functions to generate artificial positive samples by warping labeled vessels and adding them to negative samples.
   - **Pipeline Integration:** Ensure the pipeline script incorporates the augmented dataset during the training phase.

3. **Specific Changes:**
   - **Modify Training Logic:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/model/training.py`
     - **Lines:** Around 221-232
     - **Description:** Modify the DataLoader to include the augmented dataset.
     - **Code Snippet:**
       ```python
       # Assuming `augmented_dataset` is a function that returns the augmented dataset
       train_dl = torch.utils.data.DataLoader(augmented_dataset(train_imgs, train_masks, transform),
                                              batch_size=2,
                                              shuffle=True,
                                              collate_fn=custom_collate,
                                              num_workers=1,
                                              pin_memory=True if torch.cuda.is_available() else False)
       ```

   - **Add Utility Functions for Augmentation:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/utils.py`
     - **Description:** Add functions to generate artificial positive samples.
     - **Code Snippet:**
       ```python
       def augment_dataset_with_negatives_and_positives(image_names, mask_names):
           # Logic to add negative samples
           # Logic to create artificial positive samples by warping labeled vessels
           # Return the augmented dataset
           pass
       ```

   - **Integrate Augmented Dataset in Pipeline:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/pulakatlab/auto_slide/src/pipeline/run_pipeline.py`
     - **Lines:** Around 87-98
     - **Description:** Ensure the training function uses the augmented dataset.
     - **Code Snippet:**
       ```python
       def run_model_training(data_dir=None):
           logger.info("Starting model training with augmented dataset...")
           from model.training import main as training_main

           try:
               # Pass augmented dataset to training
               training_main(augmented=True)
               logger.info("Model training completed successfully")
               return True
           except Exception as e:
               logger.error(f"Error in model training: {str(e)}")
               return False
       ```

These changes aim to incorporate dataset augmentation into the training pipeline, potentially improving model performance by providing more diverse training data.

---
*This response was automatically generated by blech_bot using model gpt-4o*

---
*This response was automatically generated by blech_bot*

**Comment by abuzarmahmood** [comment]
For `Logic to create artificial positive samples by warping labeled vessels`, index and image by its mask, apply a random affine transformation, and overwrite another image with the warped, indexed vessel image

**Comment by abuzarmahmood** [comment]
Created pull request: https://github.com/pulakatlab/auto_slide/pull/17
Continue discussion there.

---
*This response was automatically generated by blech_bot*
